server/sonar-server/src/main/java/org/sonar/server/edition/StandaloneEditionManagementStateImpl.java

```diff
@@ -0,0 +1,184 @@
+    return Optional.ofNullable(pendingEditionKey);
+  }
+
+  @Override
+  public Optional<String> getPendingLicense() {
+    return Optional.ofNullable(pendingLicense);
+  }
+
+  @Override
+  public synchronized PendingStatus startAutomaticInstall(License license) {
+    checkLicense(license);
+    changeStatusFromTo(NONE, AUTOMATIC_IN_PROGRESS);
+    this.pendingLicense = license.getContent();
+    this.pendingEditionKey = license.getEditionKey();
+    persistProperties();
+    return this.pendingInstallationStatus;
+  }
+
+  @Override
+  public synchronized PendingStatus startManualInstall(License license) {
+    checkLicense(license);
+    changeStatusFromTo(NONE, MANUAL_IN_PROGRESS);
+    this.pendingLicense = license.getContent();
+    this.pendingEditionKey = license.getEditionKey();
+    this.pendingInstallationStatus = MANUAL_IN_PROGRESS;
+    persistProperties();
+    return this.pendingInstallationStatus;
+  }
+
+  @Override
+  public PendingStatus newEditionWithoutInstall(String newEditionKey) {
+    requireNonNull(newEditionKey, "newEditionKey can't be null");
+    checkArgument(!newEditionKey.isEmpty(), "newEditionKey can't be empty");
+    changeStatusFromTo(NONE, NONE);
+    this.currentEditionKey = newEditionKey;
+    persistProperties();
+    return this.pendingInstallationStatus;
+  }
+
+  @Override
+  public synchronized PendingStatus automaticInstallReady() {
+    changeStatusFromTo(AUTOMATIC_IN_PROGRESS, AUTOMATIC_READY);
+    persistProperties();
+    return this.pendingInstallationStatus;
+  }
+
+  @Override
+  public synchronized PendingStatus finalizeInstallation() {
+    checkState(this.pendingInstallationStatus == AUTOMATIC_READY || this.pendingInstallationStatus == MANUAL_IN_PROGRESS,
+      "Can't finalize installation when state is %s", this.pendingInstallationStatus);
+
+    this.pendingInstallationStatus = NONE;
+    this.currentEditionKey = this.pendingEditionKey;
+    this.pendingEditionKey = null;
+    this.pendingLicense = null;
+    persistProperties();
+    return this.pendingInstallationStatus;
+  }
+
+  private void changeStatusFromTo(PendingStatus expectedStatus, PendingStatus newStatus) {
+    checkState(pendingInstallationStatus == expectedStatus,
+      "Can't move to {} when status is {} (should be {})",
+      newStatus, pendingInstallationStatus, expectedStatus);
+    this.pendingInstallationStatus = newStatus;
+  }
+
+  private void persistProperties() {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      InternalPropertiesDao internalPropertiesDao = dbClient.internalPropertiesDao();
+      if (pendingInstallationStatus == NONE) {
+        internalPropertiesDao.saveAsEmpty(dbSession, PENDING_EDITION_KEY);
+        internalPropertiesDao.saveAsEmpty(dbSession, PENDING_LICENSE);
+      } else {
+        internalPropertiesDao.save(dbSession, PENDING_EDITION_KEY, pendingEditionKey);
+        internalPropertiesDao.save(dbSession, PENDING_LICENSE, pendingLicense);
```

Comment:
Wouldn't it be nice to always write all keys, no matter what is the situation, so that they are consistent with each other?

Reply:
this is the case